lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -120,6 +120,33 @@ static DestructureTupleInst *getSingleDestructureTupleUser(SILValue value) {
   return result;
 }
 
+/// Given a value:
+/// - If it is not tuple-typed, add the value to `results`.
+/// - If it is tuple-typed and has a single `destructure_tuple` user, collect
+///   all `destructure_tuple` results in `results`.
+/// - If it is tuple-typed and does not have a single `destructure_tuple` user,
+///   collect all `tuple_extract` users in `results`. Only extracted elements
+///   are collected; non-extracted elements are initialized in `results` as
+///   `SILValue()`.
+static void collectAllExtractedElements(SILValue value,
```

Comment:
This function should not be necessary. There are only two cases to handle: 1. `value` is a tuple, in which case you get its single `destructure_tuple` user and its result list (of type `SILInstructionResultArray`), and 2. `value` is not a tuple, in which case you use `value` directly. It's more efficient to just use `if (auto *dti = getSingleDestructureTupleUser(value)) { ... } else { ... }` at call sites because it won't create a new `SmallVector`.

Reply:
`That makes sense, I'll change call sites to use `if (auto *dti = getSingleDestructureTupleUser(value)) { ... } else { ... }`.

Reply:
Good point. After discussion, we ended up on a `forEachApplyDirectResult` helper, [done here](https://github.com/apple/swift/pull/27358/commits/503d554893e514859f583af5b2d390e604e0a1ad).